### ISSUE TITLE
Provide a convenient method for lazy Component creation based on `HashMap::entry`

### ISSUE THREAD

**Issue by viridia** [bevyengine/bevy#10635]
## What problem does this solve or what need does it fill?

I've run across a number of cases where I need to gain access to a `Component` regardless of whether or not it exists: that is, if the component exists, I want a mutable reference to it; if it does not exist, I want it to be created and return a mutable reference to the newly created component. Unfortunately, the current way of doing this is somewhat awkward:

```rust
let e = world.entity_mut(entity);
let &mut ref = match e.get_mut::<MyComponent>::() {
    Some(mut cmp) => cmp,
    None => {
        e.insert(MyComponent { ... });
        e.get_mut::<Component>().unwrap()
    }
}
```
Because `insert` returns `Self`, there's no easy way to access the component that just got inserted, other than looking it up again.

## What solution would you like?

What I'd like is some kind of `get_or_create` method that only creates the component if it doesn't already exist:

```rust
let e = world.entity_mut(entity);
let &mut ref = e.get_or_create::<MyComponent>::(|| MyComponent { ... });
```
Basically this would check to see if the component already exists: if so, then return a reference to it. Otherwise, it calls the closure to initialize a new component, and inserts that into the entity, returning a reference to the newly-inserted component.

## What alternative(s) have you considered?

For now, I'm doing it the awkward way.

**Comment by DasLixou** [comment]
Implementing some sort of `entry` api like hashmaps do might be interesting...

**Comment by alice-i-cecile** [comment]
Yeah, `entry` feels like the place to borrow from. Note that as a workaround, you can use `if let` to help with this. As you saw on Discord from @aevyrie:

```rust
if let Some(mut reactive) = world.get_mut::<Reactive<C>>(entity) {
    // ... mutable acces to the `Reactive` component in here
    // &mut world is released here because we were only using for mutable access to the component
} else {
    // ... otherwise we insert it
    world.entity_mut(entity).insert(Reactive { /* ... */ });
}
```

You could also modify this to a `let else` block.

In any case, I'm in favor of a method on `EntityMut` that does this dance for you :)

**Comment by 13ros27** [comment]
This should be fixed by #10650 which will allow you to simply use
```
world.entity_mut(entity).entry().or_insert_with(|| MyComponent { ... });
```
